### DISCUSSION TITLE
Failed to create a Webhook Definition from the Elsa Dashboard

### DISCUSSION THREAD

**Discussion by jayachandra21** [elsa-workflows/elsa-core#1522] [Q&A]
Hi,

When I am trying to create a webhook from "**Webhook definition**" tab in Elsa Dashboard, getting the error as **"An error occurred: Request failed with status code:400"**

![image](https://user-images.githubusercontent.com/67507118/133044523-9869381f-4fa3-4273-b4d9-09178efe8cdc.png)

I have used the latest Elsa packages as 2.2.1
![image](https://user-images.githubusercontent.com/67507118/133044812-2e943993-aa0f-4bbd-af7e-dcbf94bfc6cd.png)

**Comment by jayachandra21** [comment]
Hi @sfmskywalker,

As you have mentioned that above reported issue is resolved in the latest pre released packages, I have used the pre release packages and observed that "**Webhook Definition**" tab it self not available in the Elsa Dashboard.

![image](https://user-images.githubusercontent.com/67507118/133045265-16044393-8504-4702-b6f5-dbd6196a8e23.png)

Elsa Dashboard with pre released packages:

![image](https://user-images.githubusercontent.com/67507118/133045307-70485462-d2f8-4acb-bc51-d78bca0aeec4.png)

**Comment by sfmskywalker** [comment]
Hi @jayachandra1221 , I should have mentioned that the webhooks module is optional, which means you need to manually enable the corresponding plugin from HTML /JavaScript on the frontend and install the necessary services on the backend. 

The latter is straightforward:

1. Add a package reference to `Elsa.Webhooks.Api`.
2. Call `AddWebhooks` on the Elsa options builder (e.g. `services.AddElsa(elsa => elsa.AddWebhooks())`);

To install the front-end plugin, use the following code (update values as necessary, such as the server URL):

```html
<elsa-studio-root server-url="https://localhost:11000" monaco-lib-path="build/assets/js/monaco-editor/min" culture="en-US">
  <elsa-studio-dashboard></elsa-studio-dashboard>
</elsa-studio-root>

<script type="module">
  // Import publicly exposed services.
  import { WebhooksPlugin } from "/build/index.esm.js";

  // Get a handle to the elsa-studio-root element.
  const elsaStudioRoot = document.querySelector('elsa-studio-root');

  // Configure Elsa during the 'initializing' event.
  elsaStudioRoot.addEventListener('initializing', e => {
    const elsa = e.detail;
    elsa.pluginManager.registerPlugin(WebhooksPlugin);
  });

</script>
```

**Reply by jayachandra21** [reply]
I have updated my project with steps you have provided. Now I am getting the following exception:
![image](https://user-images.githubusercontent.com/67507118/133244094-b7269261-9ab2-47a2-b959-25f588fb2f47.png)

![image](https://user-images.githubusercontent.com/67507118/133244660-ee873610-a344-4b61-94ca-7b8456b26bf8.png)

Steps I have done in Project as:
1. Added the Packages reference in the solution.
![image](https://user-images.githubusercontent.com/67507118/133245109-6de8ade2-6e5b-4b92-a5e4-f2eaf78e3482.png)

2. Updated the Startup.cs class with service register
![image](https://user-images.githubusercontent.com/67507118/133245276-21166981-bf1f-41dc-a4f0-b7b691e094e9.png)

3. Updated the **Pages/_Host.cshtml** with following code snippet
![image](https://user-images.githubusercontent.com/67507118/133245498-a1db5642-98ac-4daa-9776-1d22aa17633e.png)

**Reply by sfmskywalker** [reply]
Looks like the persistence provider didn't get registered. By default, this is the InMemory provider, but after a quick glance at the code this only gets registered if you use the "features" feature. But that should be optional and there's no fluent extension method for you to use at this point. I will look into it some more.

**Reply by sfmskywalker** [reply]
I just pushed an update that fixes this, so it should be available soon from MyGet.

By default, the InMemory provider will be used, but you can change this to e.g. EF Core with SQLite as follows:

```csharp
services.AddElsa(elsa => 
{
   elsa.AddWebhooks(webhooks => webhooks.UseEntityFrameworkPersistence(ef => ef.UseWebhookSqlite()))
}
```

**Reply by jayachandra21** [reply]
I have used the latest packages from the MyGet and update the "AddWebhooks" method in Startup.cs class with UseEntityFrameworkPersistence.

Now at this time, AddWebhooks method throwing an error.

Did I miss anything ?

![image](https://user-images.githubusercontent.com/67507118/133274253-4efc625e-72b1-4097-b329-688ea3bcd43c.png)

**Reply by sfmskywalker** [reply]
What’s the error?

**Reply by sfmskywalker** [reply]
Oh wait, I see the error, sorry :)

**Reply by sfmskywalker** [reply]
Looks like you may be missing a package. Try adding the Elsa.Webhooks.Persistence.EntityFramework.SqlServer package (doing this from memory so exact package name might differ a bit).

**Reply by jayachandra21** [reply]
Now getting different issue,
![image](https://user-images.githubusercontent.com/67507118/133279481-4fcff52a-367c-4749-85f6-09343d897a5b.png)

Configured in the Startup.cs file as,
![image](https://user-images.githubusercontent.com/67507118/133279576-bdbc2852-4e89-4d5e-b092-e033f48d39b6.png)

Did I missed any configuration? registering the Wehbook definition service?

**Comment by jayachandra21** [comment]
Hi @sfmskywalker,

I have referred the latest package files in the project, but still issue appears. 
I have attached sample project for the reference.

![image](https://user-images.githubusercontent.com/67507118/133925827-7182b0d2-5d72-4f13-9468-a160f1d1a66a.png)
[WebApplication1.zip](https://github.com/elsa-workflows/elsa-core/files/7191988/WebApplication1.zip)

**Reply by sfmskywalker** [reply]
Thanks for the excellent repro! That was very helpful. As it turns out, there were a couple of issues still with the persistence registration for the Webhooks module. I just pushed a fix that should work well now. If not, please do let me know. The version with the fix should be the upcoming `2.3.0-preview-211`.

**Reply by jayachandra21** [reply]
Hi @sfmskywalker,

I have used the latest package version `2.3.0-preview-220`. However, now I am facing different kind issue as below.
![image](https://user-images.githubusercontent.com/67507118/134187292-60aa9364-6aea-4829-b150-88341e9a96ec.png)

**Reply by sfmskywalker** [reply]
I forgot to mention: in order for the webhooks migrations to run automatically, you need to specify a marker type of the assembly containing said migrations.

Look at the UseSqlServer method, you’ll see it accepts a type. If you provide a type such as `WebhookSqlServerElsaContextFactory` then it will use migrations from its assembly.

I realize it’s not obvious, but neither is a good solution to me as of yet :)

**Reply by jayachandra21** [reply]
Thank You. Now that issue got resolved :) 

But for now, could not able to see the `Webhook Definition`  button in front end. I have added the script block which you have provided in the above message trail.

```
@page "/"
@{
    var serverUrl = $"{Request.Scheme}://{Request.Host}";
}
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Elsa Workflows</title>
    <link rel="icon" type="image/png" sizes="32x32" href="/_content/Elsa.Designer.Components.Web/elsa-workflows-studio/assets/images/favicon-32x32.png">
    <link rel="icon" type="image/png" sizes="16x16" href="/_content/Elsa.Designer.Components.Web/elsa-workflows-studio/assets/images/favicon-16x16.png">
    <link rel="stylesheet" href="/_content/Elsa.Designer.Components.Web/elsa-workflows-studio/assets/fonts/inter/inter.css">
    <link rel="stylesheet" href="/_content/Elsa.Designer.Components.Web/elsa-workflows-studio/elsa-workflows-studio.css">
   
    <script src="/_content/Elsa.Designer.Components.Web/monaco-editor/min/vs/loader.js"></script>
    <script type="module" src="/_content/Elsa.Designer.Components.Web/elsa-workflows-studio/elsa-workflows-studio.esm.js"></script>
    
</head>
<body>
    <elsa-studio-root server-url="@serverUrl" monaco-lib-path="build/assets/js/monaco-editor/min" culture="en-US">
        <elsa-studio-dashboard></elsa-studio-dashboard>
    </elsa-studio-root>
    <script type="module">
        // Import publicly exposed services.
        import { WebhooksPlugin } from "/build/index.esm.js";

        // Get a handle to the elsa-studio-root element.
        const elsaStudioRoot = document.querySelector('elsa-studio-root');

        // Configure Elsa during the 'initializing' event.
        elsaStudioRoot.addEventListener('initializing', e => {
            const elsa = e.detail;
            elsa.pluginManager.registerPlugin(WebhooksPlugin);
        });
    </script>

</body>
</html>
```

**Reply by sfmskywalker** [reply]
Try replacing the path used in the import attribute with `/_content/Elsa.Designer.Components.Web/elsa-workflows-studio/elsa-workflows-studio.esm.js`

**Reply by jayachandra21** [reply]
Hi @sfmskywalker,

In brower console, system shows does not provide an export named `WebhooksPlugin`. Is it something different export name I should refer?

![image](https://user-images.githubusercontent.com/67507118/134313369-62198d98-2f9d-4846-84bf-ac7584ee4fc4.png)

**Reply by sfmskywalker** [reply]
My bad, it should be: `/_content/Elsa.Designer.Components.Web/elsa-workflows-studio/index.esm.js`.

**Reply by jayachandra21** [reply]
Hi @sfmskywalker,

Yes, it is working amazing 👍  Thank You!

**Reply by mikecasas** [reply]
@jayachandra21 Would you be willing to share your solution in either another zip or repo? I followed along the steps, but still can't see the web hook definition on the screen. 

When I change the line below to point to ```index.esm.js``` nothing renders in the browser.
```
<script type="module" src="/_content/Elsa.Designer.Components.Web/elsa-workflows-studio/elsa-workflows-studio.esm.js"></script>
```


My host file is below with the error message I receive.

```
@page "/"
@{
    var serverUrl = $"{Request.Scheme}://{Request.Host}";
}
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Elsa Workflows</title>
    <link rel="icon" type="image/png" sizes="32x32" href="/_content/Elsa.Designer.Components.Web/elsa-workflows-studio/assets/images/favicon-32x32.png">
    <link rel="icon" type="image/png" sizes="16x16" href="/_content/Elsa.Designer.Components.Web/elsa-workflows-studio/assets/images/favicon-16x16.png">
    <link rel="stylesheet" href="/_content/Elsa.Designer.Components.Web/elsa-workflows-studio/assets/fonts/inter/inter.css">
    <link rel="stylesheet" href="/_content/Elsa.Designer.Components.Web/elsa-workflows-studio/elsa-workflows-studio.css">

    <script src="/_content/Elsa.Designer.Components.Web/monaco-editor/min/vs/loader.js"></script>
    <script type="module" src="/_content/Elsa.Designer.Components.Web/elsa-workflows-studio/elsa-workflows-studio.esm.js"></script>

</head>
<body>

    <elsa-studio-root server-url="@serverUrl" monaco-lib-path="build/assets/js/monaco-editor/min" culture="en-US">
        <elsa-studio-dashboard></elsa-studio-dashboard>
    </elsa-studio-root>
    <script type="module">
        // Import publicly exposed services.
        import { WebhooksPlugin } from "/build/index.esm.js";

        // Get a handle to the elsa-studio-root element.
        const elsaStudioRoot = document.querySelector('elsa-studio-root');

        // Configure Elsa during the 'initializing' event.
        elsaStudioRoot.addEventListener('initializing', e => {
            const elsa = e.detail;
            elsa.pluginManager.registerPlugin(WebhooksPlugin);
        });
    </script>

</body>
</html>

```






![issue](https://user-images.githubusercontent.com/11261164/138352896-225ceff8-0644-44a8-ae36-a88399f02af1.PNG)

**Reply by jayachandra21** [reply]
Hi @mikecasas,

There are couple of points on Webhook defition:

- Default Webhook is not enabled in the latest packages.So you need to explicitly enable to see the same in the Dashboard.
- Packages references should be added relevant to webhook(Elsa.Webhooks.Api,Elsa.Webhooks.Persistence.EntityFramework.SqlServer) etc.,
[Startup.zip](https://github.com/elsa-workflows/elsa-core/files/7402599/Startup.zip)

**Reply by mikecasas** [reply]
Thanks so much. I got it working. My module import was wrong.
